File: 0085/solution.java
import java.util.Stack;

class Solution {
    public int maximalRectangle(char[][] matrix) {
        int ROWS = matrix.length;
        int COLS = matrix[0].length;
        int[] totals = new int[COLS];
        int area = 0;

        for (int row = 0; row < ROWS; row++) {
            for (int col = 0; col < COLS; col++) {
                if (matrix[row][col] == '1') {
                    totals[col] += 1;
                } else {
                    totals[col] = 0;
                }
            }
            area = Math.max(area, rectangleArea(totals));
        }
        return area;
    }

    public int rectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        int maxArea = 0;
        for (int i = 0; i < heights.length; i++) {
            while (!stack.isEmpty() && heights[i] < heights[stack.peek()]) {
                int tp = stack.pop();
                int width = stack.isEmpty() ? i : i - 1 - stack.peek();
                maxArea = Math.max(maxArea, heights[tp] * width);
            }
            stack.push(i);
        }
        while (!stack.isEmpty()) {
            int tp = stack.pop();
            int width = stack.isEmpty() ? heights.length : heights.length - 1 - stack.peek();
            maxArea = Math.max(maxArea, heights[tp] * width);
        }
        return maxArea;
    }
}

// time O(n*m)
// space O(m)

File: 0085/solution.ts
function maximalRectangle(matrix: string[][]): number {
  const ROWS = matrix.length;
  const COLS = matrix[0].length;
  let totals: number[] = new Array(COLS).fill(0);
  let area = 0;

  for (let row = 0; row < ROWS; row++) {
    for (let col = 0; col < COLS; col++) {
      if (matrix[row][col] === "1") {
        totals[col]++;
      } else {
        totals[col] = 0;
      }
    }
    area = Math.max(area, rectangleArea(totals));
  }
  return area;
}

function rectangleArea(heights: number[]): number {
  let stack: number[] = [];
  let maxArea = 0;
  for (let i = 0; i < heights.length; i++) {
    while (stack.length > 0 && heights[i] < heights[stack[stack.length - 1]]) {
      let height = heights[stack.pop()!];
      let width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;
      maxArea = Math.max(maxArea, height * width);
    }
    stack.push(i);
  }
  while (stack.length > 0) {
    let height = heights[stack.pop()!];
    let width =
      stack.length === 0
        ? heights.length
        : heights.length - stack[stack.length - 1] - 1;
    maxArea = Math.max(maxArea, height * width);
  }
  return maxArea;
}
